File: codeforces/16/B/solution.go
package main

import (
	"fmt"
	"math"
)

func min(nums ...int) int {
	res := math.MaxInt32
	for _, num := range nums {
		if num < res {
			res = num
		}
	}
	return res
}

func main() {
	n, m, a, b, countMap := 0, 0, 0, 0, map[int]int{}
	fmt.Scan(&n, &m)

	for i := 0; i < m; i++ {
		fmt.Scan(&a, &b)
		countMap[b] += a
	}

	pointer, result := 10, 0
	for (n > 0) && (pointer > 0) {
		steal := min(n, countMap[pointer])
		result, n, pointer = result+steal*pointer, n-steal, pointer-1
	}

	fmt.Println(result)
}

File: codeforces/16/B/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
  int n, m, a, b;
  cin >> n >> m;

  map<int, int> countMap;
  for (int i = 0; i < m; i++) {
    cin >> a >> b;
    countMap[b] += a;
  }

  int pointer = 10, result = 0;
  while ((pointer > 0) && (n > 0)) {
    int steal = countMap[pointer] > n ? n : countMap[pointer];
    result += steal * pointer, pointer -= 1, n -= steal;
  }
  cout << result << endl;
}